### PR TITLE
HPCC-23282 Correct reading of a UTF-16 file

### PR DESCRIPTION
**Pull request by AttilaVamos**
This is an experimental feature.

Add "bool keepSourceEncoding" parameter to SprayVariable(), SprayDelimited(), SprayXml() and SprayJson() functions to keep original source encoding or enable target file to be UTF8 encoded. Defaults to TRUE.

Add same to DFUPlus spray

Handle encoding and improve error logging.

Add regression test case to test keeping original encoding and converting to UTF-8.

This is a retargeted and updated version of https://github.com/hpcc-systems/HPCC-Platform/pull/13860

## Type of change:
- [ ] This change is a bug fix (non-breaking change which fixes an issue).
- [x] This change is a new feature (non-breaking change which adds functionality).
- [ ] This change improves the code (refactor or other change that does not change the functionality)
- [ ] This change fixes warnings (the fix does not alter the functionality or the generated code)
- [ ] This change is a breaking change (fix or feature that will cause existing behavior to change).
- [ ] This change alters the query API (existing queries will have to be recompiled)

## Checklist:
- [x] My code follows the code style of this project.
  - [x] My code does not create any new warnings from compiler, build system, or lint.
- [x] The commit message is properly formatted and free of typos.
  - [x] The commit message title makes sense in a changelog, by itself.
  - [x] The commit is signed.
- [x] My change requires a change to the documentation.
  - [ ] I have updated the documentation accordingly, or...
  - [ ] I have created a JIRA ticket to update the documentation.
  - [ ] Any new interfaces or exported functions are appropriately commented.
- [x] I have read the CONTRIBUTORS document.
- [x] The change has been fully tested:
  - [x] I have added tests to cover my changes.
  - [x] All new and existing tests passed.
  - [ ] I have checked that this change does not introduce memory leaks.
  - [ ] I have used Valgrind or similar tools to check for potential issues.
- [ ] I have given due consideration to all of the following potential concerns:
  - [ ] Scalability
  - [ ] Performance
  - [ ] Security
  - [ ] Thread-safety
  - [ ] Cloud-compatibility
  - [ ] Premature optimization
  - [ ] Existing deployed queries will not be broken
  - [ ] This change fixes the problem, not just the symptom
  - [ ] The target branch of this pull request is appropriate for such a change.
- [x] There are no similar instances of the same problem that should be addressed
  - [x] I have addressed them here
  - [ ] I have raised JIRA issues to address them separately
- [ ] This is a user interface / front-end modification
  - [ ] I have tested my changes in multiple modern browsers
  - [ ] The component(s) render as expected

## Smoketest:
- [ ] Send notifications about my Pull Request position in Smoketest queue.
- [ ] Test my draft Pull Request.

## Testing:

Tested manually and added a new regression test case

### CODE REVIEW

**Comment by AttilaVamos** [comment]
@jakesmith @ghalliday this is a re-targeted and updated version of https://github.com/hpcc-systems/HPCC-Platform/pull/13860. Please review.

File: dali/dfuplus/dfuplus.cpp [hunk 1]
```diff
@@ -429,6 +429,13 @@ bool CDfuPlusHelper::fixedSpray(const char* srcxml,const char* srcip,const char*
     if(globals->hasProp("expireDays"))
         req->setExpireDays(globals->getPropInt("expireDays"));
 
+    const char* encoding = globals->queryProp("encoding");
+    DFUfileformat formatCode = CDFUfileformat::decode(encoding);
+    if (formatCode != DFUff_unknown)
+        req->setSourceFormat(encoding);
+    else
+        throw MakeStringException(0, "Unknown source file format: '%s'", encoding);
```
[the remark this answers is not in the document]
Comment (jakesmith) [reply dali/dfuplus/dfuplus.cpp:437]: What sets 'encoding' if dfuplus format=fixed ?

**Comment by jakesmith** [comment]
@AttilaVamos - a few questions/comments.

File: dali/ft/filecopy.cpp [hunk 2]
```diff
@@ -2690,6 +2711,9 @@ void FileSprayer::setTarget(IDistributedFile * target)
             srcFormat.separate.set("\\,");
 
         tgtFormat.set(srcFormat);
+        if ((operation == dfu_import) && (!options->getPropBool("@keepSourceEncoding", true)))
+            tgtFormat.set(FFTutf8);
```
Comment (jakesmith) [inline dali/ft/filecopy.cpp:2715]: this doesn't sound right if source format is fixed?

**Comment by AttilaVamos** [comment]
Close because it needs redesign and re-implement.

**Pull request closed without merging**